:sparkles: Add new geocoding provider: Bing

wlorenzetti opened the pull request:
Closes: #645

Add the capability to the end user of G3W-SUITE to set the provider for '_geocoding_' map control (old '_nominatim_'). 

I the section of '_Options and actions_' of qdjango project form page, it was added a multi select box for choose the geoconding providers to use in the webgis client.

![Screenshot_20231108_152611](https://github.com/g3w-suite/g3w-admin/assets/196809/9bda47e5-387f-4d8a-b30d-3c06d2120665)

The providers can be set by the Django settings variable `GEOCONDING_PROVIDERS` who the default value is:

```python
GEOCODING_PROVIDERS = {
  "nominatim": {
    "label": "Nominatim (OSM)",
    "url": "https://nominatim.openstreetmap.org/search",
  },
  "bing_streets": {
    "label": "Bing Streets",
    "url": "https://dev.virtualearth.net/REST/v1/Locations/",
  },
  "bing_places": {
    "label": "Bing Places",
    "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/",
  },
}
```

**NB**:  `bing` providers will be available in the _Geoconding providers_ select box if a bing key it was set inside the Django settings [VENDOR_KEYS](https://g3w-suite.readthedocs.io/en/latest/settings.html?highlight=vendor#vendor-keys)

Commit:
Add GEOCODING_PROVIDERS setting.

Commit:
Add geocoding_providers to qdjango.project model.

Commit:
Update base settings GEOCODING_PROVIDERS.

Commit:
Add qdjango utility: get_geocoding_providers.

Commit:
Add geocoding_provider CRUD.

Commit:
Change initconfig API REST proeperty group.mapcontrols from list to dict.

wlorenzetti commented:
@volterra79 @Raruto a this point i suggest to rename _nominatim_ mapcontrol in _geocoding_. It'd be possible?

wlorenzetti commented:
@volterra79 @Raruto now in _initconfig_ variable (or API REST) the _mapcontrols_ property it was changed:

```js
"mapcontrols": {
            "zoombox": {},
            "zoom": {},
            "zoomhistory": {},
            "zoomtoextent": {},
            "query": {},
            "querybbox": {},
            "querybypolygon": {},
            "querybydrawpolygon": {},
            "overview": {},
            "scaleline": {},
            "geolocation": {
                "nominatim": {
                    "label": "Nominatim (OSM)",
                    "url": "https://nominatim.openstreetmap.org/search"
                },
                "bing": {
                    "label": "Bing",
                    "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/"
                }
            },
            "streetview": {},
            "nominatim": {
                "nominatim": {
                    "label": "Nominatim (OSM)",
                    "url": "https://nominatim.openstreetmap.org/search"
                },
                "bing": {
                    "label": "Bing",
                    "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/"
                }
            },
            "addlayers": {},
            "length": {},
            "area": {},
            "mouseposition": {},
            "scale": {},
            "screenshot": {},
            "geoscreenshot": {}
        }
```

Commit:
Fix for update projects with geocoding_providers empty or None.

Commit:
Check for geocoding_providers is empty.

Raruto commented:
@wlorenzetti perhaps you meant `"geocoding"` ?

```jsonc
"geolocation": { // <-- "geolocation" does not require custom providers
  "nominatim": {
      "label": "Nominatim (OSM)",
      "url": "https://nominatim.openstreetmap.org/search"
  },
  "bing": {
      "label": "Bing",
      "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/"
  }
},
```

I haven't tested it yet, but I think it's wrong within the PR code too.

Commit:
Update geoconding_providers qdjango.project field for backward compatibility.

Commit:
Fix test.

wlorenzetti commented:
> @wlorenzetti perhaps you meant `"geocoding"` ?
> 
> ```js
> "geolocation": { // <-- "geolocation" does not require custom providers
>   "nominatim": {
>       "label": "Nominatim (OSM)",
>       "url": "https://nominatim.openstreetmap.org/search"
>   },
>   "bing": {
>       "label": "Bing",
>       "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/"
>   }
> },
> ```
> 
> I haven't tested it yet, but I think it's wrong within the PR code too.

@Raruto now it 'd work fine.



> @wlorenzetti perhaps you meant `"geocoding"` ?

Yes :)

> I haven't tested it yet, but I think it's wrong within the PR code too.

@Raruto try now it'd work fine now.

Commit:
Merge branch 'dev' into Set_geocoding_provider

Commit:
Add sql to migration for change mapcontrol nominatim in geocoding

Commit:
Update mapcontrol structure

Commit:
Small fix.

Commit:
Typo.

Commit:
Add email exists check on recovery username form.

Commit:
Change from 'provider' to 'providers' inside the options of mapcontrol 'geocoding'.

Commit:
Change from 'provider' to 'providers' inside the options of mapcontrol 'geocoding'.

Commit:
Refactoring and add settings to reset password and user recovery email templates.

Raruto commented on g3w-admin/qdjango/utils/models.py:
```diff
@@ -163,4 +163,19 @@ def get_view_layer_ids(user, project):
+
+    ret = []
+    for gp, p in settings.GEOCONDING_PROVIDERS.items():
+        if gp == 'bing' and 'bing' not in settings.VENDOR_KEYS:
+            continue
```
@wlorenzetti hiding it can become difficult to use (and discover), I would recommend showing an informational message instead, eg:

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/8bcfa3a4-87a8-4421-b0de-e4841bdaf417)

```html
<div class="help-block" style="color: red">
  ⚠️ No <a href="https://g3w-suite.readthedocs.io/en/latest/settings.html#vendors-settings" target="_blank"><code>VENDOR_KEYS</code></a> found for "Bing" provider
</div>
```

Correct me if I'm wrong, only `Nominatim` provider will be activated by default for each project (so it wouldn't send invalid provider configurations to the client anyway):

![image](https://github.com/g3w-suite/g3w-admin/assets/9614886/0f5dcf04-54af-45b6-b581-f5e52133a17e)

wlorenzetti replied:
@Raruto for me is correct not show Bing if a a vendor key for it is not set inside the settings. It could be useful add a description for the find int he form writing a thing like this: "For use Bing geoding is necessary have a Bing API key "

> Correct me if I'm wrong, only Nominatim provider will be activated by default for each project (so it wouldn't send invalid provider configurations to the client anyway):

No why? at most you will have a map control without provider in the configuration API

Raruto replied:
> hiding it can become difficult to use (and discover), I would recommend showing an informational message instead
> > for me is correct not show Bing if a a vendor key for it is not set inside the settings

I said that because it complicates development a bit.

API keys (and other server configurations) can also be added on the fly via JS.

If we also have to check the docker configuration it becomes an extra step (a "duplicated" step) which would require a dedicated documentation..

Even on the server side, a sysadmin might even be happy to simply do that:

```py
GEOCONDING_PROVIDERS = {
  "bing": {
    "label": "Bing",
    "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/?key=my.super.secret.key"
  }
}
```

which is equivalent to the wordy version:

```py
VENDOR_KEYS = {
 'bing': 'my.super.secret.key'
}

GEOCONDING_PROVIDERS = {
  "bing": {
    "label": "Bing",
    "url": "https://dev.virtualearth.net/REST/v1/LocalSearch/"
  }
}
```

But without the risk to lose them in configuration files.

In essence, let's leave people a little more freedom on how to handle it.

wlorenzetti replied:
@Raruto ok I can understand what you are saying me, but the VENDORS_KEYS are used also for other features i.e. base layers .. so I think if better leave one place where to set the VENDOR key.

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505
    editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72

wlorenzetti committed:
:sparkles: Change password at first login (#650)

* Add PASSWORD_CHANGE_FIRST_LOGIN setting.

* Add change_password_first_login property to Userdata model.

* Create custom G3WSetPasswordForm and custom G3WLoginView to use for change password at first login workflow.

* Start testing.

* Testing.

* Split change password at first login from reset password workflow.

* Create a custom view for confirm reset password to split User rest password workflow from Change password at first login worflow.

* Override settings fix.

---------

Co-authored-by: wlorenzetti <lorenzett@gis3w.it>

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/commits/bf79291cac027d84d325cdb93d7d9fe1c9a798b6

wlorenzetti committed:
:sparkles: Add support to qgis `'postgresraster'` layer (#652)

* Add 'postgresraster' qgis layer type support.

* :arrow_up: Client:
    g3w-client :https://github.com/g3w-suite/g3w-client/pull/519

---------

Co-authored-by: wlorenzetti <lorenzett@gis3w.it>
Co-authored-by: volterra79 <boccacci.francesco@gmail.com>

wlorenzetti committed:
Fix get filter layer saved with anonymous user. (#654)

Co-authored-by: wlorenzetti <lorenzett@gis3w.it>

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/commits/d610c3e85547b3888c463080b423566ee10f76ba

Commit:
Typo

Commit:
Fix importing project with RelationReference widget

Commit:
Fix migrations.

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/commits/1a8c128e98ee74c77062cad858fd5c1ba189adaa

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/ Missing svg icon

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/commits/23227b6b325975c7ae50d281b71f3c9b57d260f4

volterra79 committed:
:arrow_up: Client:
    g3w-client : https://github.com/g3w-suite/g3w-client/pull/505/commits/573bc3b0afaf92864355b641a976eb064b403e2e

Commit:
Merge branch 'dev' into Set_geocoding_provider

# Conflicts:
#	g3w-admin/client/static/client/css/app.min.css
#	g3w-admin/client/static/client/js/app.min.js

volterra79 committed:
:arrow_up: Client:
    g3w-client :https://github.com/g3w-suite/g3w-client/pull/505/commits/bfd21d6e55d8c92fd310e2e2e5029c937e84efc4
    g3w-client-plugin-editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72/commits/546be8b9e0bef576810a0407432f2d62b32e8290

volterra79 committed:
:arrow_up: Client:
    g3w-client-plugin-editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72/commits/8e2a163cd356c68b897ca498f937b76954c8b526

volterra79 committed:
:arrow_up: Client:
    g3w-client-plugin-editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72/commits/bfc43de302c1e4bdc8824611fdf433852e6141c8

volterra79 committed:
:arrow_up: Client:
    g3w-client-plugin-editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72/commits/cfd558d0eb9cfea04dd83019220d897150d4bf51

volterra79 committed:
:arrow_up: Client:
    g3w-client: https://github.com/g3w-suite/g3w-client/pull/505/commits/7ac2a8c673010a0cd3226159b70edabd5331d441

Raruto committed:
mispelling

volterra79 committed:
:arrow_up: Client:
    g3w-client: https://github.com/g3w-suite/g3w-client/pull/505/commits/ad1b3c2399edb2c612ec66bf48fed458a78897a6
    g3w-client-plugin-editing: https://github.com/g3w-suite/g3w-client-plugin-editing/pull/72/commits/21cc2471e3023c9340a324b3681e2ee7c9b8f1b6

Commit:
Remove 'Bing' from default value of 'GEOCODING_PROVIDERS' setting.

volterra79 committed:
:arrow_up: Client:
    g3w-client: https://github.com/g3w-suite/g3w-client/pull/505/commits/b23276bf494781c6aed11c49f1ec67918676bd3b

volterra79 committed:
:arrow_up: Client:
    g3w-client: Add geocoding providers folder
